api/src/controllers/login.js

```diff
@@ -551,30 +542,39 @@ module.exports = {
     }
   },
   oidcLogin: async (req, res) => {
-    req.body = { locale: 'en' };
+    const limited = await rateLimitService.isLimited(req);
+    if (limited) {
+      return serverUtils.rateLimited(req, res);
+    }
     const currentUrl =  new URL(`${req.protocol}://${req.get('host')}${req.originalUrl}`);
     try {
-      const auth = await sso.getIdToken(currentUrl);
-      const cookie = await sso.getCookie(auth.user.username);
-      const redirectUrl = await setCookies(req, res, null, cookie);
+      const { preferred_username, locale } = await sso.getIdToken(currentUrl);
+      const sessionCookie = await sso.getCookie(preferred_username);
+      req.body = { locale };
+      const redirectUrl = await setCookies(req, res, sessionCookie);
       res.status(302).redirect(redirectUrl);
     } catch (e) {
       logger.error('Error logging in via SSO: %o', e);
-      return handleSsoLoginError(e, req, res);
+      const redirectUrl = url.format({
+        pathname: path.join('/', environment.db, 'login'),
+        query: { sso_error: e.status === 401 ? 'ssouserinvalid' : 'loginerror' }
+      });
+      res.status(302).redirect(redirectUrl);
     }
   },
   oidcAuthorize: async (req, res) => {
-    const redirectUrl = new URL(
-      `/${environment.db}/login/oidc/get_token`,
-      `${req.protocol}://${req.get('host')}`
-    ).toString();
+    const limited = await rateLimitService.isLimited(req);
+    if (limited) {
+      return serverUtils.rateLimited(req, res);
+    }
 
+    const redirectUrl = new URL(`${req.protocol}://${req.get('host')}/${environment.db}/login/oidc`);
     try {
-      const authUrl = await sso.getAuthorizationUrl(redirectUrl);
-      res.status(302).redirect(authUrl.href);
+      const authUrl = await sso.getAuthorizationUrl(redirectUrl.toString());
+      res.status(302).send(authUrl.href);
     } catch (e) {
       logger.error('Error getting authorization redirect url for SSO: %o', e);
-      return handleSsoLoginError(e, req, res);
+      return serverUtils.error(e, req, res);
```

Comment:
Have a look at my changes in the `public/login/script.js` file to better understand this change.  Basically, I realized that we could make this endpoint a bit more generic (and more closely mirror the normal login error behavior) by letting the login page do the redirecting based on what is returned from this endpoint.  This lets us do the error handling for the authorize call right on the login page instead of needing to re-direct back again.